<a id='toc'></a>

### Table of Contents
* [0. Identify Business needs](#business)<br>
    * [0.1. Define Key facts](#key)<br>
* [1. Import Data](#import) <br>
    * [1.1. Import the needed libraries](#lib)<br>
    * [1.2. Import and integrate data](#integrate)<br>
    * [1.3. Check for duplicates](#duplicates)<br>
* [2. Explore Data](#explore) <br>
    * [2.1. Basic Exploration](#basic)<br>
    * [2.2. Statistical Exploration](#stats)<br>
    * [2.3. Visual Exploration](#visual)<br>
    * [2.4. In-depth Exploration](#depth)<br>
* [3. Preprocess Data](#preprocess) <br>
    * [3.1. Data Cleaning](#clean)<br>
        * [3.1.1. Outliers](#outliers)<br>
        * [3.1.2. Missing Values](#missing)<br>
    * [3.2. Data Transformation](#transform)<br>
        * [3.2.1. Create new variables](#new)<br>
        * [3.2.2. Misclassifications](#misc)<br>
        * [3.2.3. Incoherencies](#inco)<br>
        * [3.2.4. Binning](#bin)<br>
        * [3.2.5. Reclassify](#rec)<br>
        * [3.2.6. Power Transform](#power)<br>
        * [3.2.7. Creating Dummy variables](#dummy)<br>
        * [3.2.8. Scaling](#scale)<br>
    * [3.3. Data Reduction](#reduce)<br>
        * [3.3.1. Multicollinearity - Check correlation](#corr)<br>
        * [3.3.2. Unary Variables](#unary)<br>
        * [3.3.3. Variables with a high percentage of missing values](#na)<br>
        * [3.3.4. PCA](#pca)

<a id='business'></a>

# <font color = '#cc9900'> 0. Identify Business needs </font>

• BookMe is well-established international company, working on the hospitalarity sector. It provides accomodations to tourists and travellers, whether for leisure or business. <br><br>
• The accomodations can only be booked online, through their website. Currently, the company has around 30,000 registered clients and serves more than 100,000 consumers a year. <br><br>
• The main goal of BookMe company is to provide the best conditions as possible, so by the end of the stay, each customer receives a survey to evaulate their satisfaction level, over the different services provided: location, proce, amenities, among others. The scale rate varies between 0 and 5. <br><br>
• The profit growth perspectives for the next three years are fickle, so the marketing team needs a new efficiency program to improve their activities and boost their campaigns, in order to invert the situation. <br>

__Cluster.csv__

| Attribute | Description | 
| --- | --- |
| Name | Customer’s name |
| Year_Birth | Customer’s birth year |
| Longevity | Whether the customer registered more than 1 year ago or not (yes or no) |
| Churn | Whether the customer churned or not (churn or nochurn) |
| TypeTravel | Customer’s reason for travelling (business or leisure) |
| RoomType | Type of room reserved |
| RewardPoints |  Customer’s rewarding point for loyalty |
| Comfort | Satisfaction level of customer regarding comfort of the room (0 to 5) |
| ReceptionSchedule | Satisfaction level of customer regarding reception schedule (0 to 5) |
| FoodDrink | Satisfaction level of customer regarding food and drink available (0 to 5) |
| Location | Satisfaction level of customer regarding accommodation location (0 to 5) |
| Wifi | Satisfaction level of customer regarding wi-fi service (0 to 5) |
| Amenities | Satisfaction level of customer regarding accommodation amenities (0 to 5) |
| Staff | Satisfaction level of customer regarding accommodation amenities (0 to 5) |
| OnlineBooking | Satisfaction level of customer regarding online booking ease (0 to 5) |
| PriceQuality | Satisfaction level of customer regarding price quality relationship (0 to 5) |
| RoomSpace | Satisfaction level of customer regarding room space (0 to 5) |
| CheckOut | Satisfaction level of customer regarding check-out (0 to 5) |
| CheckIn | Satisfaction level of customer regarding check-in (0 to 5) |
| Cleanliness |  Satisfaction level of customer regarding cleanliness (0 to 5) |
| BarService |  Satisfaction level of customer regarding bar service (0 to 5) |

<a id='key'></a>

### 0.1. Define Key facts

__`Step 0`__ Detail the key facts for our project:

**Name**: BookMe <br><br>
**Sector**: Hospitality<br><br>
**Business focus**: Providing rooms with the best conditions possible<br><br>
**Business facts**<br>
International website<br>
30.000 registered customers<br>
100.000 consumers/year<br>

**Case facts**<br>
Survey is sent to customer at the end of the stay<br>
Rating scale 0 to 5<br>
Rate location, price, amenities, provided and others<br><br>
**Issue**<br>
Stable revenues and healthy bottom line in past 3 years _but_ profit growth perspectives for next 3 years are fickle<br><br>
**Business challenge**<br>
Marketing wants to improve marketing activities, focusing on boosting the marketing campaigns' efficiency<br><br>
**Goal**<br>
Clustering the different types of customers that the company has to create more efficient campaigns<br><br>

**Requirements**
* Identify actionable segments;
    * Must only use the current dataset;
* Two visions must be developed:
    * Customer satisfactions;
    * Customer characteristics;
    * Other perspectives are optional;
    
**Deliverables**
1. Report that identifies the main customer segments;
    * Max. 8 pages
    * File name "202122 Cluster GroupXX Notebook.ipynb"
        * Heading 1: Arial, Size 12 pt, in bold
        * Heading 2 (if needed): Arial, Size 11 pt, in bold and italic
        * Text: Arial, Size 10 pt, line space of 1.5 points.
        * Margins: The default ones in word (Top, Bottom, Left and Right as 1”)
    * All the figures and tables should be included in the Annexes (not included in the 8 pages mentioned previously)

2. First draft of a marketing plan.

<a id='lib'></a>

### 1.1. Import the needed libraries

__`Step 1`__ Import the following libraries/functions: 
    - pandas as pd 
    - pyplot from matplotlib as plt 
    - seaborn as sns
    - MinMaxScaler from sklearn.preprocessing

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import MinMaxScaler

<a id='integrate'></a>

### 1.2. Import and integrate data

__`Step 2`__ Import the excel file `cluster.csv` and store it in the object `demo` <br>

In [14]:
df = pd.read_csv('cluster.csv')
df

Churn                    Name Longevity  Year_Birth TypeTravel  \
0        churn       Ms. Nicole Clarke       yes      1974.0   business   
1      nochurn         Mr. Jesse Scott       yes      1965.0   business   
2        churn  Mr. Christopher Carter       yes      1973.0   business   
3      nochurn     Ms. Jennifer Morgan       yes      1993.0    leisure   
4      nochurn       Mr. Michael White       yes      1989.0   business   
...        ...                     ...       ...         ...        ...   
15584    churn       Ms. Wendy Roberts        no      1988.0   business   
15585    churn         Ms. Sonya Smith        no      2000.0   business   
15586  nochurn      Mr. Maxwell Wright       yes      1964.0   business   
15587  nochurn       Mr. Gregory Moran       yes      1992.0    leisure   
15588    churn       Mr. Steven Mendez       yes      1983.0   business   

      RoomType  RewardPoints  Comfort  ReceptionSchedule  FoodDrink  ...  \
0       single          4907        3                  4          1  ...   
1       single          6724        1                  1          1  ...   
2       single          4365        3                  3          3  ...   
3       double          3849        1                  1          1  ...   
4       single          5376        2                  2          3  ...   
...        ...           ...      ...                ...        ...  ...   
15584   double          4968        2                  0          2  ...   
15585   double          4970        3                  3          3  ...   
15586   single          4046        1                  1          1  ...   
15587   double          5155        4                  3          4  ...   
15588    suite          5136        3                  3          3  ...   

       Wifi  Amenities  Staff  OnlineBooking  PriceQuality  RoomSpace  \
0         4          3      4              3             3          3   
1         5          3      4              5             5          5   
2         1          4      4              3             3          2   
3         4          4      5              4             4          4   
4         5          5      5              5             5          3   
...     ...        ...    ...            ...           ...        ...   
15584     3          2      4              3             4          5   
15585     5          3      5              5             3          1   
15586     3          4      5              4             4          4   
15587     5          4      5              5             2          5   
15588     3          3      3              3             4          5   

       CheckOut  Checkin  Cleanliness  BarService  
0             3        4            3           4  
1             5        1            5           2  
2             3        2            3           1  
3             4        4            4           3  
4             4        1            3           5  
...         ...      ...          ...         ...  
15584         1        3            4           3  
15585         3        4            4           5  
15586         4        5            4           4  
15587         4        3            4           5  
15588         4        1            3           3  

[15589 rows x 21 columns]

<a id='duplicates'></a>

### 1.3. Check for duplicates

__`Step 3`__ Drop any duplicate rows present in the dataframe with the method `drop_duplicates()

In [15]:
df[df.duplicated()]

Churn                  Name Longevity  Year_Birth TypeTravel RoomType  \
8195  nochurn      Ms. Abigail York       yes      1995.0    leisure   double   
9176    churn   Ms. Abigail Kennedy       yes      1991.0   business    suite   
9417  nochurn  Ms. Abigail Buchanan       yes      1972.0   business   double   

      RewardPoints  Comfort  ReceptionSchedule  FoodDrink  ...  Wifi  \
8195          5098        5                  5          5  ...     4   
9176          5932        3                  3          2  ...     3   
9417          6769        5                  4          4  ...     5   

      Amenities  Staff  OnlineBooking  PriceQuality  RoomSpace  CheckOut  \
8195          5      5              3             3          4         3   
9176          3      3              3             4          1         4   
9417          5      4              5             5          5         5   

      Checkin  Cleanliness  BarService  
8195        3            3           5  
9176        3            4           3  
9417        2            5           1  

[3 rows x 21 columns]

In [16]:
df.drop_duplicates(inplace = True)
df

Churn                    Name Longevity  Year_Birth TypeTravel  \
0        churn       Ms. Nicole Clarke       yes      1974.0   business   
1      nochurn         Mr. Jesse Scott       yes      1965.0   business   
2        churn  Mr. Christopher Carter       yes      1973.0   business   
3      nochurn     Ms. Jennifer Morgan       yes      1993.0    leisure   
4      nochurn       Mr. Michael White       yes      1989.0   business   
...        ...                     ...       ...         ...        ...   
15584    churn       Ms. Wendy Roberts        no      1988.0   business   
15585    churn         Ms. Sonya Smith        no      2000.0   business   
15586  nochurn      Mr. Maxwell Wright       yes      1964.0   business   
15587  nochurn       Mr. Gregory Moran       yes      1992.0    leisure   
15588    churn       Mr. Steven Mendez       yes      1983.0   business   

      RoomType  RewardPoints  Comfort  ReceptionSchedule  FoodDrink  ...  \
0       single          4907        3                  4          1  ...   
1       single          6724        1                  1          1  ...   
2       single          4365        3                  3          3  ...   
3       double          3849        1                  1          1  ...   
4       single          5376        2                  2          3  ...   
...        ...           ...      ...                ...        ...  ...   
15584   double          4968        2                  0          2  ...   
15585   double          4970        3                  3          3  ...   
15586   single          4046        1                  1          1  ...   
15587   double          5155        4                  3          4  ...   
15588    suite          5136        3                  3          3  ...   

       Wifi  Amenities  Staff  OnlineBooking  PriceQuality  RoomSpace  \
0         4          3      4              3             3          3   
1         5          3      4              5             5          5   
2         1          4      4              3             3          2   
3         4          4      5              4             4          4   
4         5          5      5              5             5          3   
...     ...        ...    ...            ...           ...        ...   
15584     3          2      4              3             4          5   
15585     5          3      5              5             3          1   
15586     3          4      5              4             4          4   
15587     5          4      5              5             2          5   
15588     3          3      3              3             4          5   

       CheckOut  Checkin  Cleanliness  BarService  
0             3        4            3           4  
1             5        1            5           2  
2             3        2            3           1  
3             4        4            4           3  
4             4        1            3           5  
...         ...      ...          ...         ...  
15584         1        3            4           3  
15585         3        4            4           5  
15586         4        5            4           4  
15587         4        3            4           5  
15588         4        1            3           3  

[15586 rows x 21 columns]

[BACK TO TOC](#toc)

<a id='explore'></a>

# <font color = '#cc9900'> 2. Explore Data </font>

<a id='basic'></a>

### 2.1. Basic Exploration

__`Step 4`__ Check the number of rows and columns in the dataset using the attribute `shape`

In [17]:
df.shape

(15586, 21)

Our data set contains 15,5586 rows and 21 columns.

__`Step 5`__ Check the name of the columns of our dataset using the attribute `columns`

In [18]:
df.columns

Index(['Churn', 'Name', 'Longevity', 'Year_Birth', 'TypeTravel', 'RoomType',
       'RewardPoints', 'Comfort', 'ReceptionSchedule', 'FoodDrink', 'Location',
       'Wifi', 'Amenities', 'Staff', 'OnlineBooking', 'PriceQuality',
       'RoomSpace', 'CheckOut', 'Checkin', 'Cleanliness', 'BarService'],
      dtype='object')

__`Step 6`__ Check the first three rows of the dataset using the method `head()` and the last three rows using the method `tail()`

In [20]:
df.head(3)

Churn                    Name Longevity  Year_Birth TypeTravel RoomType  \
0    churn       Ms. Nicole Clarke       yes      1974.0   business   single   
1  nochurn         Mr. Jesse Scott       yes      1965.0   business   single   
2    churn  Mr. Christopher Carter       yes      1973.0   business   single   

   RewardPoints  Comfort  ReceptionSchedule  FoodDrink  ...  Wifi  Amenities  \
0          4907        3                  4          1  ...     4          3   
1          6724        1                  1          1  ...     5          3   
2          4365        3                  3          3  ...     1          4   

   Staff  OnlineBooking  PriceQuality  RoomSpace  CheckOut  Checkin  \
0      4              3             3          3         3        4   
1      4              5             5          5         5        1   
2      4              3             3          2         3        2   

   Cleanliness  BarService  
0            3           4  
1            5           2  
2            3           1  

[3 rows x 21 columns]

In [21]:
df.tail(3)

Churn                Name Longevity  Year_Birth TypeTravel RoomType  \
15586  nochurn  Mr. Maxwell Wright       yes      1964.0   business   single   
15587  nochurn   Mr. Gregory Moran       yes      1992.0    leisure   double   
15588    churn   Mr. Steven Mendez       yes      1983.0   business    suite   

       RewardPoints  Comfort  ReceptionSchedule  FoodDrink  ...  Wifi  \
15586          4046        1                  1          1  ...     3   
15587          5155        4                  3          4  ...     5   
15588          5136        3                  3          3  ...     3   

       Amenities  Staff  OnlineBooking  PriceQuality  RoomSpace  CheckOut  \
15586          4      5              4             4          4         4   
15587          4      5              5             2          5         4   
15588          3      3              3             4          5         4   

       Checkin  Cleanliness  BarService  
15586        5            4           4  
15587        3            4           5  
15588        1            3           3  

[3 rows x 21 columns]

__`Step 7`__ Get more information of the dataset by calling the method `info()`

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15586 entries, 0 to 15588
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Churn              15586 non-null  object 
 1   Name               15586 non-null  object 
 2   Longevity          15586 non-null  object 
 3   Year_Birth         15391 non-null  float64
 4   TypeTravel         15586 non-null  object 
 5   RoomType           15586 non-null  object 
 6   RewardPoints       15586 non-null  int64  
 7   Comfort            15586 non-null  int64  
 8   ReceptionSchedule  15586 non-null  int64  
 9   FoodDrink          15586 non-null  int64  
 10  Location           15586 non-null  int64  
 11  Wifi               15586 non-null  int64  
 12  Amenities          15586 non-null  int64  
 13  Staff              15586 non-null  int64  
 14  OnlineBooking      15586 non-null  int64  
 15  PriceQuality       15586 non-null  int64  
 16  RoomSpace          155

We can verify that we are working with:
- 1 float variable
- 15 integer variables
- 5 object variables

We can also check that some of the variables have missing values (Year_Birth). We are going to deal with this in a further step.

<a id='stats'></a>

### 2.2. Statistical Exploration

#### 2.3.1.) Numerical Variables

__`Step 8`__ Get the main descriptive statistics for all the numeric variables in using the method `.describe()`

In [24]:
df.describe().T

count         mean          std     min     25%     50%  \
Year_Birth         15391.0  1981.705607    15.179757  1936.0  1970.0  1981.0   
RewardPoints       15586.0  5022.418581  1027.940126   409.0  4445.0  5088.0   
Comfort            15586.0     2.841332     1.388541     0.0     2.0     3.0   
ReceptionSchedule  15586.0     2.997049     1.519035     0.0     2.0     3.0   
FoodDrink          15586.0     2.844412     1.436937     0.0     2.0     3.0   
Location           15586.0     2.985821     1.299438     1.0     2.0     3.0   
Wifi               15586.0     3.244963     1.327064     0.0     2.0     3.0   
Amenities          15586.0     3.374631     1.352419     0.0     2.0     4.0   
Staff              15586.0     3.506288     1.319625     1.0     3.0     4.0   
OnlineBooking      15586.0     3.454190     1.310401     0.0     2.0     4.0   
PriceQuality       15586.0     3.459579     1.268180     1.0     3.0     4.0   
RoomSpace          15586.0     3.470871     1.293782     0.0     2.0     4.0   
CheckOut           15586.0     3.700500     1.158693     1.0     3.0     4.0   
Checkin            15586.0     3.327409     1.266944     1.0     3.0     3.0   
Cleanliness        15586.0     3.692288     1.154484     1.0     3.0     4.0   
BarService         15586.0     3.347427     1.300371     0.0     2.0     3.0   

                       75%     max  
Year_Birth         1994.00  2014.0  
RewardPoints       5648.75  6950.0  
Comfort               4.00     5.0  
ReceptionSchedule     4.00     5.0  
FoodDrink             4.00     5.0  
Location              4.00     5.0  
Wifi                  4.00     6.0  
Amenities             4.00     5.0  
Staff                 5.00     5.0  
OnlineBooking         5.00     5.0  
PriceQuality          4.00     5.0  
RoomSpace             5.00     5.0  
CheckOut              5.00     5.0  
Checkin               4.00     5.0  
Cleanliness           5.00     5.0  
BarService            4.00     5.0

From the above table, we can get some conclusions. Some examples are:
- `count`- The Year_Birth variable has 15391 valid values. We have a problem of missing values in here.
- `mean`- In average, my customers satisfaction level with the majority all the services provided varies between 3 and 4.
- `std`- The standard deviation of RewardPoints is quite high. This indicates that the values are spread out over a wider range.
- `min`- Various services were rated with level 0, except for Location, Staff, PriceQuality, CheckOut, CheckIn and Cleanliness.
- `50%`- Half of my customers rated with levels 3 and 4 all the services provided. Half of our customer have rated with level 4 some crucial services: Amenities, Staff, OnlineBooking, PriceQuality, RoomSpace, CheckOut and Cleanliness.
- `max`- The maximum rate value for Wifi is 6. This is an incoherence - according to the business needs, the range is between 1 and 5. As we can see, all services have been rated with level 5, at least once.